rezasm-source/rezasm-core/src/instructions/instruction.rs

```diff
@@ -32,17 +36,37 @@ impl Instruction {
     pub fn get_function(&self) -> &TInstructionFunction {
         &self.function
     }
+
+    pub fn get_name(&self) -> &String {
+        &self.name
+    }
+
+    pub fn call_function(
+        &self,
+        simulator: &mut Simulator,
+        arguments: &Vec<ArgumentType>,
+    ) -> Result<(), SimulatorError> {
+        (self.function)(simulator, &self.types, arguments)
+    }
 }
 
-pub fn matches_argument_types(target: &Vec<TypeId>, attempt: &Vec<ArgumentType>) -> bool {
-    if target.len() != attempt.len() {
-        false
-    } else {
-        for (t, a) in zip(target, attempt) {
-            if t != &a.get_mut_type_id() && t != &a.get_ref_type_id() && t != &a.get_type_id() {
-                return false;
-            }
+#[macro_export]
+macro_rules! instruction {
+    ($name:ident, |$simulator_name:ident: Simulator, $($names:ident: $types:ty),*| $func:tt) =>
+    ({
+        let mut v: Vec<std::any::TypeId> = Vec::new();
+        $(v.push(std::any::TypeId::of::<&mut $types>());)*
+        fn $name($simulator_name: &mut crate::simulation::simulator::Simulator, types: &Vec<std::any::TypeId>, arguments: &Vec<crate::instructions::argument_type::ArgumentType>) -> Result<(), crate::util::error::SimulatorError> {
+            let mut _counter: usize = 0;
+            $(
+                #[allow(unused_mut)]
+                let mut $names: $types = arguments[_counter].clone().into();
```

Comment:
You will need to change this if you implement the TryInto trait instead of the Into trait.